### PR TITLE
[FSDP] feat: Add FSDP forward pefetch and recompute chunking entropy

### PR DESCRIPTION
**Pull request by CurryRice233**
### Checklist Before Starting

- [x] Search for similar PR(s).

### What does this PR do?

1. Add fsdp1 forward pefetch configuration.
2. Add chunk entropy computation.
3. Add torch.checkpoint to entropy computation.
4. Move data to device from `ActorRolloutRefWorker.update_actor` to `DataParallelPPOActor.update_policy`.
5. Add `npu_cross_entropy_loss` fusion kernel.

### High-Level Design

1. More detail see [FSDP forward_pefetch](https://docs.pytorch.org/docs/stable/fsdp.html#module-torch.distributed.fsdp)
2. `logits` usually is a large tensor [bsz\*seq_len, voc], on `compute_entropy_from_logits` will use [bsz\*seq_len, voc] * (4(float32) + 2(autocast of softmax+logsumexp) + 1(output of softmax)) memory. To reduce this memory peak, we can use chunk calculation, changing [bsz*seq_len, voc] to [chunk_size(2048), voc].
3. During the training phase, `enable_gradient_checkpointing=True` is not applicable to entropy calculation, so add the recomputation function of entropy to reduce the memory peak during the training phase.
4. On `ActorRolloutRefWorker.update_actor` all batch data is moved to the device, but this is unnecessary, `DataParallelPPOActor.update_policy` will move the data to the device for each micro batch.

### Specific Changes

> List the specific changes.

### API

Add 3 new configurations in actor/ref, 1 new configuration in critic/reward.

- actor_rollout_ref.actor.fsdp_config.forward_prefetch: False
- actor_rollout_ref.actor.entropy_from_logits_with_chunking: False
- actor_rollout_ref.actor.entropy_checkpointing: False
- actor_rollout_ref.ref.fsdp_config.forward_prefetch: False
- actor_rollout_ref.ref.entropy_from_logits_with_chunking: False
- actor_rollout_ref.ref.entropy_checkpointing: False
- critic.model.fsdp_config.forward_prefetch: False
- reward_model.model.fsdp_config.forward_prefetch: False

### Usage Example

> Provide usage example(s) for easier usage.

```python
# Add code snippet or script demonstrating how to use this 
```

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluatuion results, etc.

### Additional Info.

- **Issue Number**: Fixes issue # or discussion # if any.
- **Training**: [Note which backend this PR will affect: FSDP, Megatron, both, or none]
- **Inference**: [Note which backend this PR will affect: vLLM, SGLang, both, or none]

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [x] Add `[BREAKING]` to the PR title if it breaks any API.
- [x] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [x] New CI unit test(s) are added to cover the code path.
- [x] Rely on existing unit tests on CI that covers the code path.

### CODE REVIEW

File: verl/trainer/config/ppo_trainer.yaml [hunk 1]
```diff
@@ -201,6 +201,12 @@ actor_rollout_ref:
     # Sequence parallelism size for Ulysses-style model parallelism
     ulysses_sequence_parallel_size: 1
 
+    # calculate entropy with chunking to reduce memory peak
+    entropy_from_logits_with_chunking: False
```
Comment (vermouth1992) [inline verl/trainer/config/ppo_trainer.yaml:205]: Can we remove these two options? I guess they should be on by default for NPU

Comment (CurryRice233) [reply verl/trainer/config/ppo_trainer.yaml:205]: > Can we remove these two options? I guess they should be on by default for NPU

not only for NPU, but also for GPU, chunk+recompute can reduce entropy memory utilization(e2e memory peak). If memory is sufficient, its recommended to disable these two options.

**Comment by eric-haibin-lin** [comment]
@CurryRice233 thx for the contribution! Would you mind adding these options to https://verl.readthedocs.io/en/latest/perf/perf_tuning.html (https://github.com/volcengine/verl/blob/main/docs/perf/perf_tuning.rst) ?

**Comment by CurryRice233** [comment]
> @CurryRice233 thx for the contribution! Would you mind adding these options to https://verl.readthedocs.io/en/latest/perf/perf_tuning.html (https://github.com/volcengine/verl/blob/main/docs/perf/perf_tuning.rst) ?

No problem, I will create a PR within this week.
